src/lib/units/offset.js

```diff
@@ -103,40 +108,42 @@ hooks.updateOffset = function () {};
 // _changeInProgress == true case, then we have to adjust, because
 // there is no such time in the given timezone.
 export function getSetOffset (input, keepLocalTime, keepMinutes) {
-    var offset = this._offset || 0,
+    var mom = this,
+        offset = mom._offset || 0,
         localAdjust;
-    if (!this.isValid()) {
-        return input != null ? this : NaN;
+    if (!mom.isValid()) {
+        return input != null ? mom : NaN;
     }
     if (input != null) {
         if (typeof input === 'string') {
             input = offsetFromString(matchShortOffset, input);
             if (input === null) {
-                return this;
+                return mom;
             }
         } else if (Math.abs(input) < 16 && !keepMinutes) {
             input = input * 60;
         }
-        if (!this._isUTC && keepLocalTime) {
-            localAdjust = getDateOffset(this);
+        if (!mom._isUTC && keepLocalTime) {
+            localAdjust = getDateOffset(mom);
         }
-        this._offset = input;
-        this._isUTC = true;
+        mom = new Moment(mom);
+        mom._offset = input;
+        mom._isUTC = true;
         if (localAdjust != null) {
-            this.add(localAdjust, 'm');
+            mom = mom.add(localAdjust, 'm');
         }
         if (offset !== input) {
-            if (!keepLocalTime || this._changeInProgress) {
-                addSubtract(this, createDuration(input - offset, 'm'), 1, false);
-            } else if (!this._changeInProgress) {
-                this._changeInProgress = true;
-                hooks.updateOffset(this, true);
-                this._changeInProgress = null;
+            if (!keepLocalTime || mom._changeInProgress) {
+                mom = addSubtract(mom, createDuration(input - offset, 'm'), 1, false);
+            } else if (!mom._changeInProgress) {
+                mom._changeInProgress = true;
```

Comment:
This _changeInProgress was there because of mutability. Now I don't think its necessary. Or maybe it is necessary, but we should put it on the new moment, so it want fall into a cycle. Maybe that is related to the other piece of cycle-break-check in the constructor.